soc/xtensa/intel_adsp/common/CMakeLists.txt

```diff
@@ -117,10 +117,30 @@ endif()
 
 
 #  west sign
+
+# Warning: most of this code is for now duplicated in
+# soc/xtensa/nxp_adsp/CMakeLists.txt
+
 add_custom_target(zephyr.ri ALL
   DEPENDS ${CMAKE_BINARY_DIR}/zephyr/zephyr.ri
 )
 
+# Copy/normalize any weird ${CONFIG_KERNEL_BIN_NAME}.elf to zephyr.elf
+# if different.
+#
+# zephyr.elf is not used by sign.py when a boot.mod bootloader is used
+# BUT others do expect a zephyr.elf file: smex, debuggers, CI and other
+# automation, etc. (Ab)using Kconfig to merely rename a file seems like
+# a bad idea because it breaks build directory expectations.
+if(NOT CONFIG_KERNEL_BIN_NAME STREQUAL "zephyr")
```

Comment:
why the copying? why not just sign what elf name is being generated instead of hardcording to zephyr.elf?

Reply:
> why the copying? why not just sign what elf name is being generated instead of hardcording to zephyr.elf?

Because that would imply a whole lot of change in `sign.py`, `sof/scripts/xtensa-build-zephyr.py` and others for no SOF value.

Even beyond SOF I don't see the point of `CONFIG_KERNEL_BIN_NAME`; can someone share a use case?  A build directory is only a staging area and a starting point for installation / deployment / analysis / etc. so using configurable artefact filenames there is just adding a lot of complication for no obvious benefit.

It looks like CONFIG_KERNEL_BIN_NAME was added a long time ago to change the default name without breaking build directory backwards compatibility too much? It seems defined in only 3 `zephyr/**/*.conf` files right now and by only 2 persons.

Reply:
> Because that would imply a whole lot of change in `sign.py`, `sof/scripts/xtensa-build-zephyr.py` and others for no SOF value.

this is the change needed, and it is consistent with how other signers work, the binanry name should not be hardcoded and all of this need to work with whatever the application has decided to call its binary, this is the change needed for sign.py:

```diff
diff --git a/scripts/west_commands/sign.py b/scripts/west_commands/sign.py
index e57b8800b4..cffe798308 100644
--- a/scripts/west_commands/sign.py
+++ b/scripts/west_commands/sign.py
@@ -427,20 +427,22 @@ class RimageSigner(Signer):

         # warning: RIMAGE_TARGET is a duplicate of CONFIG_RIMAGE_SIGNING_SCHEMA
         target = cache.get('RIMAGE_TARGET')
+        kernel_name = build_conf.get('CONFIG_KERNEL_BIN_NAME', 'zephyr')
+
         if not target:
             log.die('rimage target not defined')

         if target in ('imx8', 'imx8m'):
-            kernel = str(b / 'zephyr' / 'zephyr.elf')
-            out_bin = str(b / 'zephyr' / 'zephyr.ri')
-            out_xman = str(b / 'zephyr' / 'zephyr.ri.xman')
-            out_tmp = str(b / 'zephyr' / 'zephyr.rix')
+            kernel = str(b / 'zephyr' / f'{kernel_name}.elf')
+            out_bin = str(b / 'zephyr' / f'{kernel_name}.ri')
+            out_xman = str(b / 'zephyr' / f'{kernel_name}.ri.xman')
+            out_tmp = str(b / 'zephyr' / f'{kernel_name}.rix')
         else:
             bootloader = str(b / 'zephyr' / 'boot.mod')
             kernel = str(b / 'zephyr' / 'main.mod')
-            out_bin = str(b / 'zephyr' / 'zephyr.ri')
-            out_xman = str(b / 'zephyr' / 'zephyr.ri.xman')
-            out_tmp = str(b / 'zephyr' / 'zephyr.rix')
+            out_bin = str(b / 'zephyr' / f'{kernel_name}.ri')
+            out_xman = str(b / 'zephyr' / f'{kernel_name}.ri.xman')
+            out_tmp = str(b / 'zephyr' / f'{kernel_name}.rix')

         # Clean any stale output. This is especially important when using --if-tool-available
         # (but not just)
```
and the make files need to be adapted as well.

> Even beyond SOF I don't see the point of `CONFIG_KERNEL_BIN_NAME`; can someone share a use case?

An application should be able to generate the binary using whatever name it likes, this has been like this for a long time.

Reply:
> this is the change needed, and it is consistent with how other signers work,

I agree this should work and fix the Zephyr build issue found by @iuliana-prodan . However it wouldn't help with various SOF scripts and automation that also pick up $NAME.lst, $NAME.map and maybe others, I mean it would not make the SOF infrastructure compatible `CONFIG_KERNEL_BIN_NAME`. As long as no one in SOF has the bad idea of using `CONFIG_KERNEL_BIN_NAME` everything should be fine.

> and the make files need to be adapted as well.

Not sure what these are.

> An application should be able to generate the binary using whatever name it likes, this has been like this for a long time.

For what purpose? If building multiple apps then `build_app1/zephyr.*`, `build_app2/zephyr.*`,... is much easier to use than `build1/app1.*`, `build2/app2.*`

This made me think about `--sysbuild`. I just tried it for the first time, using the example at
https://docs.zephyrproject.org/latest/build/sysbuild/index.html#configuration-namespacing. Out of the box it unsurprisingly does the former:
```
build/hello_world/zephyr/zephyr.elf
build/hello_world/zephyr/zephyr.map
build/hello_world/zephyr/zephyr.hex
build/hello_world/zephyr/zephyr.bin
build/hello_world/zephyr/zephyr.lst
build/hello_world/zephyr/zephyr.stat
build/hello_world/zephyr/zephyr.signed.bin
build/hello_world/zephyr/zephyr.signed.hex

build/mcuboot/zephyr/zephyr.elf
build/mcuboot/zephyr/zephyr.map
build/mcuboot/zephyr/zephyr.hex
build/mcuboot/zephyr/zephyr.bin
build/mcuboot/zephyr/zephyr.lst
build/mcuboot/zephyr/zephyr.stat

```

Reply:
@marc-hb I've tested @nashif suggestion and for NXP platforms works perfectly.
We need one more change:
```
diff --git a/soc/xtensa/nxp_adsp/CMakeLists.txt b/soc/xtensa/nxp_adsp/CMakeLists.txt
index 2ca5d100db..3586d87ffb 100644
--- a/soc/xtensa/nxp_adsp/CMakeLists.txt
+++ b/soc/xtensa/nxp_adsp/CMakeLists.txt
@@ -16,5 +16,5 @@ add_custom_command(
   OUTPUT ${CMAKE_BINARY_DIR}/zephyr/zephyr.ri
   COMMENT "west sign --if-tool-available --tool rimage ..."
   COMMAND  west sign --if-tool-available --tool rimage --build-dir ${CMAKE_BINARY_DIR}
-  DEPENDS ${CMAKE_BINARY_DIR}/zephyr/zephyr.elf
+  DEPENDS ${CMAKE_BINARY_DIR}/zephyr/${KERNEL_ELF_NAME}
 )

```

@marc-hb let me know if you update this PR, otherwise I can do it.

Thanks!

Reply:
@marc-hb I've created #60121 
Please have a look.